daphne/src/roles/helper.rs

```diff
@@ -44,254 +44,272 @@ pub trait DapHelper<S>: DapAggregator<S> {
+        self.audit_log().on_aggregation_job(
+            req.host(),
+            task_id,
+            task_config,
+            agg_job_init_req.report_shares.len() as u64,
+            AggregationJobAuditAction::Init,
+        );
 
-                // Check whether the DAP version in the request matches the task config.
-                if task_config.version != req.version {
-                    return Err(DapAbort::version_mismatch(req.version, task_config.version));
-                }
+        metrics.agg_job_started_inc();
+        metrics.inbound_req_inc(DaphneRequestType::Aggregate);
+        Ok(DapResponse {
+            version: req.version,
+            media_type: DapMediaType::AggregationJobResp,
+            payload: agg_job_resp.get_encoded(),
+        })
+    }
 
-                // Ensure we know which batch the request pertains to.
-                check_part_batch(
-                    task_id,
-                    task_config,
-                    &agg_job_init_req.part_batch_sel,
-                    &agg_job_init_req.agg_param,
-                )?;
-
-                let transition = task_config
-                    .vdaf
-                    .handle_agg_job_init_req(
-                        self,
-                        self,
-                        task_id,
-                        task_config,
-                        &agg_job_init_req,
-                        &metrics,
-                    )
-                    .map_err(DapError::Abort)
+    async fn handle_agg_job_cont_req<'req>(
+        &self,
+        req: &'req DapRequest<S>,
+        metrics: ContextualizedDaphneMetrics<'req>,
+        task_id: &TaskId,
+    ) -> Result<DapResponse, DapAbort> {
+        resolve_taskprov(self, task_id, req, None).await?;
+        let wrapped_task_config = self
+            .get_task_config_for(Cow::Borrowed(task_id))
+            .await?
+            .ok_or(DapAbort::UnrecognizedTask)?;
+        let task_config = wrapped_task_config.as_ref();
+
+        if let Some(reason) = self.unauthorized_reason(task_config, req).await? {
+            error!("aborted unauthorized collect request: {reason}");
+            return Err(DapAbort::UnauthorizedRequest {
+                detail: reason,
+                task_id: task_id.clone(),
+            });
+        }
+
+        // Check whether the DAP version in the request matches the task config.
+        if task_config.version != req.version {
+            return Err(DapAbort::version_mismatch(req.version, task_config.version));
+        }
+
+        let agg_job_cont_req =
+            AggregationJobContinueReq::get_decoded_with_param(&req.version, &req.payload)
+                .map_err(|e| DapAbort::from_codec_error(e, task_id.clone()))?;
+
+        // draft02 compatibility: In draft02, the aggregation job ID is parsed from the
+        // HTTP request payload; in the latest, the aggregation job ID is parsed from the
+        // request path.
+        let agg_job_id = match (
```

Comment:
Seems like some of this complex logic is duplicated and should be consolidated between init & cont?